compiler/rustc_lint/src/unused.rs

```diff
@@ -1153,14 +1152,35 @@ declare_lint_pass!(UnusedAllocation => [UNUSED_ALLOCATION]);
 
 impl<'tcx> LateLintPass<'tcx> for UnusedAllocation {
     fn check_expr(&mut self, cx: &LateContext<'_>, e: &hir::Expr<'_>) {
-        match e.kind {
-            hir::ExprKind::Box(_) => {}
+        let span = match e.kind {
+            hir::ExprKind::Box(_) => {
+                // Ideally, we'd underline the `box` part of the expression here,
+                // but at this point we have lost the span of the `box` keyword.
+                // Constructing a span from the inner and the entire expression's
+                // span won't work in many cases, so we just return the entire
+                // span of the `box foo`.
+                e.span
+            }
+            hir::ExprKind::Call(ref callee, _) => {
+                // Look for Box::new(foo)
+                if let hir::ExprKind::Path(ref qpath) = callee.kind &&
+                    // `Res::Local` indicates a closure
+                    let Res::Def(DefKind::Fn | DefKind::AssocFn, def_id) = cx.qpath_res(qpath, callee.hir_id) &&
```

Comment:
```suggestion
                    let Res::Def(_, def_id) = cx.qpath_res(qpath, callee.hir_id) &&
```
Unnecessary condition, `def_id` determines def kind entirely.